modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Tracker.scala

```diff
@@ -35,40 +30,37 @@ import emitters.TEmitter
  * @param namespace Tracker namespace
  * @param appId ID of the application
  * @param encodeBase64 Whether to encode JSONs
+ * @param metadata optionally a json containing the metadata context for the running instance
  */
-class Tracker(emitters: Seq[TEmitter], namespace: String, appId: String, encodeBase64: Boolean = true) {
+final case class Tracker[F[_]: Applicative](emitters: List[Emitter[F]],
+                                            namespace: String,
+                                            appId: String,
+                                            subject: Subject                     = Subject(),
+                                            encodeBase64: Boolean                = true,
+                                            metadata: Option[SelfDescribingJson] = None) {
   import Tracker._
 
-  private var subject: Subject = new Subject()
-  private var attachEc2Context = false
-  private val ec2Context       = Promise[SelfDescribingJson]
```

Comment:
My reasoning for the metadata changes:

1. The tracker won't be running on ec2 and gce at the same time, hence one field.
2. Option instead of booleans
3. Inverted the control so that metadata is now passed as a constructor parameter instead of using `Promise`. It makes `Tracker` RT and allows us to release metadata module separately as `Tracker` doesn't know about metadata providers.
4. Since with `Promise` the metadata was requested only once per `Tracker` lifetime anyway, we can skip all the `Promise<->Future` dance.